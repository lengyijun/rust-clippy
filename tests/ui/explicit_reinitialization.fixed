#![allow(dead_code)]
#![warn(clippy::explicit_reinitialization)]

fn test_copy() {
    let mut x = 1;
    println!("{x}");
    let mut x = 2;
    println!("{x}");
}

fn test_move() {
    let mut x = String::new();
    println!("{x}");
    let mut x = String::new();
    println!("{x}");
}

#[allow(unused_assignments, clippy::misrefactored_assign_op)]
fn should_lint() {
    let mut a = 1;
    let mut a = a * a * a;
}

#[allow(clippy::ptr_arg)]
fn should_lint_2() {
    fn foo(_x: &mut String) {}
    let mut s = String::new();
    let mut s = s.replacen('o', "a", 3);
    foo(&mut s);
    drop(s);
}

// require liveness analysis
#[allow(clippy::ptr_arg, clippy::unnecessary_literal_unwrap, clippy::useless_format)]
fn should_lint_3() {
    fn foo(_x: &mut String) {}
    if true {
        let mut s = Result::<String, ()>::Ok(String::new()).unwrap_or_else(|_| panic!());
        let mod_decl = format!("\nmod");
        let mut s = s.replacen(&mod_decl, "a", 3);
        foo(&mut s);
    }
    println!("123");
}

fn should_not_lint_1() {
    let mut x = 1;
    loop {
        println!("{x}");
        x = 2;
    }
}

fn should_not_lint_2() {
    let mut x = String::new();
    loop {
        println!("{x}");
        x = String::new();
    }
}

fn should_not_lint_3() {
    let mut x = String::new();
    if true {
        x = String::new();
    } else {
        x = x.trim().to_owned();
    }
    drop(x)
}

fn should_not_lint_4() {
    let mut b = true;
    loop {
        b = !b;
    }
}

fn should_not_lint_5(mut v: Vec<String>) -> Vec<String> {
    for _ in 0..10 {
        if true {
            v = Vec::new();
            continue;
        }
        v.push(String::new());
    }
    v
}

fn known_false_negative() {
    let mut x = 1;
    println!("{x}");
    {
        let mut x = 2;
    }
    println!("{x}");
}

fn false_positive() {
    let mut x = 1;
    println!("{x}");

    if true {
        x = 2;
        println!("{x}");
    } else {
        x = 3;
        println!("{x}");
    }
}

fn main() {}
