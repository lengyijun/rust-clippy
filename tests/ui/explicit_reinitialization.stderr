error: create a fresh variable is more explicit
  --> $DIR/explicit_reinitialization.rs:7:5
   |
LL |     x = 2;
   |     ^^^^^^ help: create a fresh variable instead of reinitialization: `let x = 2;`
   |
   = note: `-D clippy::explicit-reinitialization` implied by `-D warnings`
   = help: to override `-D warnings` add `#[allow(clippy::explicit_reinitialization)]`

error: create a fresh variable is more explicit
  --> $DIR/explicit_reinitialization.rs:14:5
   |
LL |     x = String::new();
   |     ^^^^^^^^^^^^^^^^^^ help: create a fresh variable instead of reinitialization: `let x = String::new();`

error: create a fresh variable is more explicit
  --> $DIR/explicit_reinitialization.rs:21:5
   |
LL |     a = a * a * a;
   |     ^^^^^^^^^^^^^^ help: create a fresh variable instead of reinitialization: `let a = a * a * a;`

error: create a fresh variable is more explicit
  --> $DIR/explicit_reinitialization.rs:28:5
   |
LL |     s = s.replacen('o', "a", 3);
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: create a fresh variable instead of reinitialization: `let mut s = s.replacen('o', "a", 3);`

error: create a fresh variable is more explicit
  --> $DIR/explicit_reinitialization.rs:40:9
   |
LL |         s = s.replacen(&mod_decl, "a", 3);
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: create a fresh variable instead of reinitialization: `let mut s = s.replacen(&mod_decl, "a", 3);`

error: create a fresh variable is more explicit
  --> $DIR/explicit_reinitialization.rs:94:9
   |
LL |         x = 2;
   |         ^^^^^^ help: create a fresh variable instead of reinitialization: `let x = 2;`

error: aborting due to 6 previous errors

